sdk/azcore/runtime/policy_bearer_token.go

```diff
@@ -42,23 +46,71 @@ func acquire(state acquiringResourceState) (newResource azcore.AccessToken, newE
 // scopes: the list of permission scopes required for the token.
 // opts: optional settings. Pass nil to accept default values; this is the same as passing a zero-value options.
 func NewBearerTokenPolicy(cred azcore.TokenCredential, scopes []string, opts *policy.BearerTokenOptions) *BearerTokenPolicy {
+	if opts == nil {
+		opts = &policy.BearerTokenOptions{}
+	}
 	return &BearerTokenPolicy{
+		authzHandler: opts.AuthorizationHandler,
 		cred:         cred,
 		scopes:       scopes,
 		mainResource: temporal.NewResource(acquire),
 	}
 }
 
+// authenticateAndAuthorize returns a function which authorizes req with a token from the policy's credential
+func (b *BearerTokenPolicy) authenticateAndAuthorize(req *policy.Request) func(policy.TokenRequestOptions) error {
+	return func(tro policy.TokenRequestOptions) error {
+		as := acquiringResourceState{p: b, req: req, tro: tro}
+		tk, err := b.mainResource.Get(as)
+		if err != nil {
+			return err
+		}
+		req.Raw().Header.Set(shared.HeaderAuthorization, shared.BearerTokenPrefix+tk.Token)
+		return nil
+	}
+}
+
 // Do authorizes a request with a bearer token
 func (b *BearerTokenPolicy) Do(req *policy.Request) (*http.Response, error) {
-	as := acquiringResourceState{
-		p:   b,
-		req: req,
+	var err error
+	if b.authzHandler.OnRequest != nil {
+		err = b.authzHandler.OnRequest(req, b.authenticateAndAuthorize(req))
+	} else {
+		err = b.authenticateAndAuthorize(req)(policy.TokenRequestOptions{Scopes: b.scopes})
 	}
-	tk, err := b.mainResource.Get(as)
+	if err != nil {
+		return nil, ensureNonRetriable(err)
```

Comment:
I think this call, and the call on line 98, need to be moved such that they only are call when the `On*()` methods are called yes?

Reply:
Well, it's an interesting question. I wrote it like this because I expect we want all errors from this policy to be nonretriable, and it seems like an oversight that we don't already wrap them to be sure. In practice, it usually doesn't matter because any errors originate in `azidentity`, whose auth errors are nonretriable. But if a user gives the policy a custom credential implementation?

Reply:
I don't think we have enough context at this point though to tell if the error is transient, do we?  e.g. if one of these extension methods fails to make an HTTP request due to some transient condition, we should ensure that the retry policy will kick in.

Reply:
If an extension method needs to send a request, it brings its own transport and is therefore responsible for retrying that request. This is baked into the common case: `azidentity` credentials have their own pipelines and retry policies.

Reply:
Oh yes of course, makes sense.